osfoffline/polling_osf_manager/polling.py

```diff
@@ -29,42 +31,84 @@ def __init__(self, user, loop):
         self.user = user
 
         self._loop = loop
+        self.poll_job = None
+        self.process_job = None
         self.osf_query = OSFQuery(loop=self._loop, oauth_token=self.user.oauth_token)
-        self.polling_event_queue = PollingEventQueue(loop=self._loop)
 
     def stop(self):
+        logger.info('OSF polling requested to stop.')
+
+        # Stop was called before start could complete
+        if not self.poll_job or not self.process_job:
+            return
+
+        # If these futures are complete an exception has been raised in them
+        if self.poll_job.done():
+            raise self.poll_job.exception()
+        if self.process_job.done():
+            raise self.process_job.exception()
 
-        self._keep_running = False
+        # Other wise cancel them, the event loop will clean up for us
+        self.poll_job.cancel()
+        self.process_job.cancel()
         self.osf_query.close()
+        logger.info('OSF polling requested stopped.')
+
+    def handle_exception(self, future):
+        # Note: The actual futures never exit, if they do an exception is raised
+        try:
+            raise future.exception()
+        except (aiohttp.ClientError, asyncio.TimeoutError):
+            logger.error('Unable to connect to the internet')
+            AlertHandler.warn('Unable to connection to OSF, we\'ll try again later.')
+        except asyncio.CancelledError:
+            # Cancellations just mean this thread is exiting
+            return
+
+        # Make sure all our jobs are cancelled
+        if future == self.poll_job:
+            self.process_job.cancel()
+        else:
+            self.poll_job.cancel()
+
+        logger.info('Restarting polling in 5 seconds')
+        # Finally restart our watcher thread in 5 seconds
+        self._loop.call_later(5, self.start)
 
     def start(self):
-        # annoying and weird way to get the remote user from the coroutine
+        remote_user = self._loop.run_until_complete(self.get_remote_user())
 
-        future = asyncio.Future(loop=self._loop)
-        asyncio.async(self.get_remote_user(future), loop=self._loop)
-        self._loop.run_until_complete(future)
-        remote_user = future.result()
+        self.queue = asyncio.Queue(maxsize=15)
 
-        self._loop.call_soon(
-            asyncio.async,
-            self.check_osf(remote_user)
-        )
+        self.process_job = asyncio.ensure_future(self.process_queue())
+        self.poll_job = asyncio.ensure_future(self.check_osf(remote_user))
+
+        self.poll_job.add_done_callback(self.handle_exception)
+        self.process_job.add_done_callback(self.handle_exception)
```

Comment:
Yes, any exception raised in the `process_queue` or `check_osf` will be handled by `handle_exception`